ci/test_python_common.sh

```diff
@@ -4,6 +4,7 @@
 set -euo pipefail
 
 . /opt/conda/etc/profile.d/conda.sh
+source ./ci/use_conda_packages_from_prs.sh
```

Comment:
Not specific to this line, but just choosing a place to start a threaded conversation.

All conda-based tests are failing with issues like this:

```text
Fatal Python error: Aborted
Thread 0x00007f1903eea640 (most recent call first):
  File "/opt/conda/envs/test/lib/python3.12/site-packages/execnet/gateway_base.py", line 534 in read
  File "/opt/conda/envs/test/lib/python3.12/site-packages/execnet/gateway_base.py", line 567 in from_io
  File "/opt/conda/envs/test/lib/python3.12/site-packages/execnet/gateway_base.py", line 1160 in _thread_receiver
  File "/opt/conda/envs/test/lib/python3.12/site-packages/execnet/gateway_base.py", line 341 in run
  File "/opt/conda/envs/test/lib/python3.12/site-packages/execnet/gateway_base.py", line 411 in _perform_spawn
Current thread 0x00007f1904b0e740 (most recent call first):
  File "/opt/conda/envs/test/lib/python3.12/site-packages/cuml/internals/api_decorators.py", line 190 in wrapper
...
Extension modules: cython.cimports.libc.math, Cython.Utils, numpy._core._multiarray_umath, numpy._core._multiarray_tests
...
.[gw13] node down: Not properly terminated
```

([build link](https://github.com/rapidsai/cuml/actions/runs/10996928514/job/30534049930?pr=6071)) 

I'll investigate this right now. I suspect it might be something related to symbol visibility (like https://github.com/rapidsai/cudf/pull/15483#discussion_r1669369852).

Reply:
Ah!! It looks like either I missed some dependencies in the test scripts, or something else is holding back the `fmt` / `spdlog` version?

I see that RAPIDS libraries are coming from `rapidsai-nightly` and not the CI artifacts, and that the older versions of `fmt` and `spdlog` are being used.

```text
...
cudf                      24.10.00a369    cuda11_py310_240923_g9b4c4c721c_369    rapidsai-nightly
cuml                      24.10.00a70     cuda11_py310_240923_ga1d76d4da_70    file:///tmp/python_channel
...
distributed-ucxx          0.40.00a        py3.10_240923_ga07a40f_28    rapidsai-nightly
...
fmt                       10.2.1               h00ab1b0_0    conda-forge
...
```

([build link](https://github.com/rapidsai/cuml/actions/runs/10996928514/job/30534056266?pr=6071))

Will try replicating that conda solve locally.

Reply:
Testing locally, I found that if I added constraints like `fmt>=11.0.2, spdlog>=1.14.1` in the conda environment, the solver was able to solve with all of `cuml`'s runtime dependencies... including the versions of `cudf`, `rmm`, `ucxx`, etc. produced from the PRs for https://github.com/rapidsai/build-planning/issues/56.

I think that `libcuml` conda packages need `host:` dependencies on `spdlog` and `fmt`.

* commit adding that: https://github.com/rapidsai/cuml/pull/6071/commits/5b812f0e966f2d8998ef62437cdcf01139103d8d
* build link to watch: https://github.com/rapidsai/cuml/actions/runs/10998530224/job/30536663624?pr=6071

That library's code directly uses both of those:

https://github.com/rapidsai/cuml/blob/7de88314d8d37ff91ac7c0ed7127487fcf0ce5c4/cpp/src/common/logger.cpp#L20-L21

https://github.com/rapidsai/cuml/blob/7de88314d8d37ff91ac7c0ed7127487fcf0ce5c4/cpp/include/cuml/common/callbackSink.hpp#L47

And having those dependencies would prevent these situations during development where the solver is able to choose to fall back to older RAPIDS nightlies from within the same release (but supporting different `fmt` / `spdlog`).

`cudf` is doing the same thing: https://github.com/rapidsai/cudf/blob/9b4c4c721c399bae9e88733da79daa1a10644481/conda/recipes/libcudf/meta.yaml#L69-L71

Reply:
This is probably the right solution, thanks for investigating.

Reply:
This ended up not being enough 😭 

Look at https://github.com/rapidsai/cuml/actions/runs/10998530224/job/30541086357?pr=6071.

The environment is still solving with `fmt==10.2.1`, `spdlog==1.12.0`, and older RAPIDS `24.10.*` nightlies that support those ([build link](https://github.com/rapidsai/cuml/actions/runs/10998530224/job/30541086357?pr=6071#step:7:925))... even though it used the `cuml` / `libcuml` packages from CI here.

```text
cudf                      24.10.00a371    cuda11_py310_240923_g625590686f_371    rapidsai-nightly
cuml                      24.10.00a71     cuda11_py310_240923_g5b812f0e9_71    file:///tmp/python_channel
...
fmt                       10.2.1               h00ab1b0_0    conda-forge
...
libcudf                   24.10.00a371    cuda11_240923_g625590686f_371    rapidsai-nightly
...
libcuml                   24.10.00a71     cuda11_240923_g5b812f0e9_71    file:///tmp/cpp_channel
libcumlprims              24.10.00a       cuda11_240923_gddfb9dd_8    rapidsai-nightly
...
rmm                       24.10.00a40     cuda11_py310_240923_g58039d3c_40    rapidsai-nightly
...
spdlog                    1.12.0               hd2e6256_2    conda-forge
...
```

I just pushed another commit explicitly pinning `fmt` and `spdlog` in the conda test environments, to force the solver to use the newer packages.

* commit: https://github.com/rapidsai/cuml/pull/6071/commits/77241136ad60d59ef462a8c506bdfd8bf216e388
* build link: https://github.com/rapidsai/cuml/actions/runs/11000193616/job/30542188379?pr=6071

This should be safe to do... as soon as there are `librmm` packages published to `rapids-nightly` containing the changes from https://github.com/rapidsai/rmm/pull/1678, every environment containing the latest `librmm` / `rmm` will have those same pins in it.

If that works, that's evidence that the `fmt` / `spdlog` update is safe and we can start moving forward.

I think both of those changes are worth merging here

- **adding the `host:` dependencies** = to make CPM unnecessary for `spdlog` and avoid unnecessary vendoring and therefore file clobbering
- **explicitly pinning in test environments** = make such fallbacks during development less likely the next time we do an update like this

Reply:
🎉 I see tests now passing and `fmt==11.0.2` + `spdlog==1.14.1` getting installed!

Build link: https://github.com/rapidsai/cuml/actions/runs/11000193616/job/30544094053?pr=6071

Another thing that I think happened here.... it's been several days since the `cudf` CI artifacts being pulled in here were produced, so the `rapidsai-nightly` channel's versions are now newer.

e.g. https://github.com/rapidsai/cudf/pull/16806 produces `cudf==24.10.00a364` but `rapidsai-nightly` now contains `cudf==24.10.00a371`. That would also explain why some of the CI artifacts are being ignored, in favor of those from `rapidsai-nightly`.

Reply:
See here: https://github.com/rapidsai/cuml/pull/6071#discussion_r1772018427